### PR TITLE
Issue130: Created A Simple RF Messaging Application (Hermes)

### PR DESCRIPTION
**Pull request by kb1lqd**
Hermes is a Python application that implements a simple messaging application allowing wireless transmission of multi-packet messages (data/text) between two Faraday digital radios. The program implements the Faraday Command program functionality that transmits a data packet directly to a remote Faraday. This program is only meant to provide an extremely simple example and functionality of transferring data between two Faraday units and is a stepping stone to improved versions/variants.

**Related Issue:** #130 

**General Functionality**
* A python based server provides Flask interfaces allowing transmission and reception from multiple locally connected Faraday units
* Example transmit/receive applications provided to demonstrate user interfacing
* Data transmissions are not guaranteed
  * Data can be lost/corrupted during reception
* Data is allowed to be up to 256 packets long (fragmentation) which is 38 Bytes per packet (9728 Bytes)

**Scripts**
* `hermesmessage.py` - A collection of functions that provide underlying functionality used to implement data transfer such as packet creation, fragmentation, state machines, etc...
* `hermesobject.py` - Python class objects that contain both transmit and receive functionality from `hermesmessage.py` and allow creation "hermes" messaging objects related to units connected to proxy.
* `hermesflaskserver.py` - A Flask server providing a basic transmit/receive API
* `hermesflaskexample_rx.py` - An example receiver program (user interface program)
* `hermesflaskexample_tx.py` - An example transmit program (user interface program)

**Example**

Left - Hermes Flask Server
Center - Hermes Receiver Example
Right - Hermes Transmitter Example

image

### CODE REVIEW

**Comment by kb1lqd** [comment]
@rtucker @reillyeon @hdkmike @kawfey may be interested to view / comment as well.

File: Applications/Hermes/Readme.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Hermes - Message Application
+
+Hermes is a simple messaging application developed for Faraday. Hermes was the God of messengers so it seemed fitting.
```
Comment (kb1lqc) [inline Applications/Hermes/Readme.md]: Second sentence and beyond: "Hermes, the messenger of the gods in mythology, is a fitting name for our first application to transition into messaging between radios."

File: Applications/Hermes/Readme.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+The messaging functionality is the result of a Flask server that allows Transmit/Receive capability for locally connected Faraday devices. When the Flask server is running, requests sent to the specific port of operation allow the transmission and reception of data messages wirelessly between two (or more) Faraday devices.
```
Comment (kb1lqc) [inline Applications/Hermes/Readme.md]: Based on Flask like many other Faraday applications, Hermes provides transmit and receive messaging functionality for Faraday radios within range of each other. Sending a message to the Flask port with an HTTP request will cause Hermes to interface with the local Faraday radio via Proxy to send the message over RF. The reverse is also true for receive.

File: Applications/Hermes/Readme.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+```Python
+[FLASK]
+HOST=127.0.0.1
+PORT=8005
+
+[PROXY]
+UNITS=2
+
+[UNIT0]
+CALLSIGN=kb1lqd
```
Comment (kb1lqc) [inline Applications/Hermes/Readme.md]: Lowercase callsign?

File: Applications/Hermes/Readme.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+NODEID=1
+
+[UNIT1]
+CALLSIGN=kb1lqd
+NODEID=2
+
+
```
Comment (kb1lqc) [inline Applications/Hermes/Readme.md]: Whitespace?

File: Applications/Hermes/Readme.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+* UNIT0
+  * Update `CALLSIGN` Replace ```REPLACEME``` to match the callsign of the Faraday unit **as assigned** in proxy
+  * Update `NODEID` to match the callsign node ID of the Faraday unit **as assigned** in proxy
+* UNIT1
+  * Update `CALLSIGN` Replace ```REPLACEME``` to match the callsign of the remote Faraday unit as configured in the devices FLASH memory configuration
+  * Update `NODEID` to match the callsign of the remote Faraday unit as configured in the devices FLASH memory configuration
+* Save the file as `hermes.ini`
+
+
+```python
```
Comment (kb1lqc) [inline Applications/Hermes/Readme.md:33]: Why is this here again? why not just have config file explained once?

File: Applications/Hermes/Readme.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+[UNIT1]
+CALLSIGN=REPLACEME
+NODEID=REPLACEME
+
+```
+
+## Run
+
+* Run the proxy server with both units properly connected to
```
Comment (kb1lqc) [inline Applications/Hermes/Readme.md]: This sentence ends abruptly

File: Applications/Hermes/hermesflaskserver.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+import hermesobject as hermesobject
+import ConfigParser
+import json
+import base64
+import cPickle
+
+from flask import Flask
+from flask import request
+
+# Open configuration INI
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py:11]: **Non-Blocking**
My Telemetry application shows [how to import safely](https://github.com/kb1lqc/Faraday-Software/blob/c26a906c62d08d484ef6804687d08dee798ae940/Applications/Telemetry/telemetry.py#L42)

File: Applications/Hermes/hermesflaskserver.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+config.read(filename)
+
+# Global variables
+dictmsgobj = {}
+
+
+def configparse():
+    """ Parses configuration file for callsigns and node ID and returns them in JSON format"""
+    local = {}
+    num = int(config.get('PROXY', 'UNITS'))
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: config.getint()

File: Applications/Hermes/hermesflaskserver.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    units = range(0, num)
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: range(num), you could also just forgo using num at all and directly use the config call.

File: Applications/Hermes/hermesflaskserver.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+    for unit in units:
+        # TODO We don't really check for valid input here yet
+        item = "UNIT" + str(unit)
+        callsign = config.get(item, "callsign").upper()
+        nodeid = config.get(item, "nodeid")
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: getint()?

File: Applications/Hermes/hermesflaskserver.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+        local[str(item)] = \
+            {
+                "callsign": callsign,
+                "nodeid": nodeid,
+            }
+
+    local = json.dumps(local)
+    return json.loads(local)
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: wait, why dumps local then loads local? you're back with the same thing in the end? This just returns a dictionary... Actually a list of dictionarie(s) I believe.

File: Applications/Hermes/hermesflaskserver.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+    Transmits a provided payload data from a specified local unit to a specific remote unit.
+
+    :param localcallsign: Flask parameter - Local unit callsign to transmit from
+    :param localnodeid: Flask parameter - Local unit node ID to transmit from
+    :param destinationcallsign: Flask parameter - Remote unit callsign to transmit to
+    :param destinationnodeid: Flask parameter - Remote unit node ID to transmit to
+    :param data: Flask parameter: Data payload to transmit to remote unit
+
+    :return: Returns JSON success status for Flask server after transmission
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py:70]: HTTP status

File: Applications/Hermes/hermesflaskserver.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+    :param localcallsign: Flask parameter - Local unit callsign to transmit from
+    :param localnodeid: Flask parameter - Local unit node ID to transmit from
+
+    :return: Returns JSON element containing the next received data from the receiver queue
+    """
+    # Global variables
+    global dictmsgobj
+
+    # If POST
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: obvious

File: Applications/Hermes/hermesflaskserver.py [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+    if request.method == 'POST':
+        # Parse Flask arguments
+        localcallsign = request.args.get("localcallsign").upper()
+        localnodeid = request.args.get("localnodeid")
+        destinationcallsign = request.args.get("destinationcallsign").upper()
+        destinationnodeid = request.args.get("destinationnodeid")
+        data = request.args.get("data")
+
+        # Get hermes TX/RX object from global dictionary
+        unitmsgobj = dictmsgobj[localcallsign + '-' + localnodeid]
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: What happens if this isn't true, the callsign+id isn't in the dictionary?

File: Applications/Hermes/hermesflaskserver.py [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+
+    Returns a JSON element containing the next received data from the receiver queue of the specified local unit.
+
+    :param localcallsign: Flask parameter - Local unit callsign to transmit from
+    :param localnodeid: Flask parameter - Local unit node ID to transmit from
+
+    :return: Returns JSON element containing the next received data from the receiver queue
+    """
+    # Global variables
+    global dictmsgobj
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: Already global in main()

File: Applications/Hermes/hermesflaskserver.py [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+def main():
+    """Main function which starts the Flask server."""
+    # Global variables
+    global dictmsgobj
+
+    # Get INI file configuration for Flask server
+    config_host = config.get("FLASK", "HOST")
+    config_port = config.get("FLASK", "PORT")
+
+    # Start the flask server on localhost:8001
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: don't use config_host... just use hermeshost = config.get(). Same for port.

File: Applications/Hermes/hermesflaskserver.py [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+    hermeshost = config_host
+    hermesport = config_port
+
+    units = configparse()
+
+    # Parse and save unit callsign-nodeid into global dictionary as keys to hermes objects
+    for key in units:
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: Key isn't key, it's probably a list of dictionaries

File: Applications/Hermes/hermesflaskserver.py [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+        unitcallsign = units[key]['callsign']
+        unitnodeid = units[key]['nodeid']
+        unitname = unitcallsign + '-' + unitnodeid
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: I actually have a create nodename function in my new aprslib branch of APRS... maybe we should combine these basic functions for Faraday into a module...

https://github.com/kb1lqc/Faraday-Software/blob/aprslib/Applications/APRS/aprs.py#L221

File: Applications/Hermes/hermesflaskserver.py (hunk 14 above)
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: [nodename function](https://github.com/kb1lqc/Faraday-Software/blob/aprslib/Applications/APRS/aprs.py#L221)?

File: Applications/Hermes/hermesflaskserver.py [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+        # Get hermes TX/RX object from global dictionary
+        unitmsgobj = dictmsgobj[localcallsign + '-' + localnodeid]
+
+        # Transmit data to remote device
+        unitmsgobj.transmit.send(destinationcallsign, int(destinationnodeid), str(data))
+
+        # Return status
+        return json.dumps(
+            {"status": "Posted Packet(s)"
+             }), 200
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: weird line split?

File: Applications/Hermes/hermesflaskserver.py [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+
+    # If GET
+    else:
+        # Parse Flask arguments
+        localcallsign = request.args.get("localcallsign").upper()
+        localnodeid = request.args.get("localnodeid")
+
+        # Get hermes TX/RX object from global dictionary
+        unitmsgobj = dictmsgobj[localcallsign + '-' + localnodeid]
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: [Nodename](https://github.com/kb1lqc/Faraday-Software/blob/aprslib/Applications/APRS/aprs.py#L221)?

File: Applications/Hermes/hermesflaskserver.py [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+
+        # Get next message from RX queue
+        received_item = unitmsgobj.receive.getqueueitem()
+
+        # Pickle dictionary data
+        received_data_pickle = cPickle.dumps(received_item)
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: Don't pickle.. you already have JSON available. Much better, more secure...

Faster: https://konstantin.blog/2010/pickle-vs-json-which-is-faster/
Secure: 
image

File: Applications/Hermes/hermesflaskserver.py [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+
+        # Encode to BASE64
+        received_item_64 = base64.b64encode(received_data_pickle)
+
+        return json.dumps(received_item_64, indent=1), 200, \
+            {'Content-Type': 'application/json'}
+
+
+@app.route('/queue', methods=['GET'])
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskserver.py]: how about /len?

File: Applications/Hermes/hermesobject.py [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+
+    :param local_device_callsign: The callsign of the proxy assigned local Faraday device to connect to
+    :param local_device_node_id: The callsign ID number of the proxy assigned local Faraday device to connect to
+    """
+    def __init__(self, local_device_callsign, local_device_node_id):
+        self.local_device_callsign = local_device_callsign
+        self.local_device_node_id = local_device_node_id
+        # Create messaging application objects needed for transmissions
+        self.faraday_tx_msg_sm = hermesmessage.MsgStateMachineTx()
+        self.faraday_tx_msg_object = hermesmessage.MessageAppTx(self.local_device_callsign,
```
Comment (kb1lqc) [inline Applications/Hermes/hermesobject.py:22]: long line

File: Applications/Hermes/hermesobject.py [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+                                                                self.local_device_node_id)
+        # Create receiver application object
+        self.faraday_rx_msg_object = hermesmessage.MessageAppRx()
+        self.rx_uart_service_port_application_number = 3
```
Comment (kb1lqc) [inline Applications/Hermes/hermesobject.py]: Config files? Or at least able to overwrite port and timeout of class instantiation?

File: Applications/Hermes/hermesobject.py [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+        return
+
+    def run(self):
+        """
+            This is the main threaded loop that the receiver program runs to obtain new received data packets from
+            the proxy interface. Received data packets are saved in to FIFO for later retrieval.
+            """
+        # Loop continuously through the faraday experimental RF command message application RX routine
+        while 1:
+            rx_message_dict = self.faraday_rx_msg_object.rxmsgloop(self.local_device_callsign,
```
Comment (kb1lqc) [inline Applications/Hermes/hermesobject.py]: Long line

File: Applications/Hermes/hermesobject.py [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+        self.fifo = Queue.Queue(0)
+        return
+
+    def run(self):
+        """
+            This is the main threaded loop that the receiver program runs to obtain new received data packets from
+            the proxy interface. Received data packets are saved in to FIFO for later retrieval.
+            """
+        # Loop continuously through the faraday experimental RF command message application RX routine
+        while 1:
```
Comment (kb1lqc) [inline Applications/Hermes/hermesobject.py:74]: No sleep to slow down processing power?

File: Applications/Hermes/hermesflaskexample_rx.py [hunk 27]
```diff
@@ -0,0 +1,53 @@
+import requests
+import base64
+import cPickle
+import time
+import ConfigParser
+import os
+
+# Open configuration INI
+config = ConfigParser.RawConfigParser()
+filename = os.path.abspath("hermes.ini")
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskexample_rx.py:10]: Same comments as hermesflaskserver.py config file import

File: Applications/Hermes/hermesflaskexample_rx.py [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+config.read(filename)
+
+localcallsign = config.get('UNIT1', 'CALLSIGN')
+localnodeid = int(config.get('UNIT1', 'NODEID'))
+
+
+def getrxqueuesize(callsign, nodeid):
```
Comment (kb1lqc) [inline Applications/Hermes/hermesflaskexample_rx.py:17]: Two items:

- I don't agree with using cpickle when JSON is already being used.
- This seems appropriate for use in the hermes module, not defined in the application sample file?

**Review by kb1lqc — DISMISSED** [review]

File: Applications/Hermes/Readme.md (hunk 1 above)
Comment (kb1lqd) [reply Applications/Hermes/Readme.md]: Fixed.

File: Applications/Hermes/Readme.md (hunk 2 above)
Comment (kb1lqd) [reply Applications/Hermes/Readme.md]: Fixed.

File: Applications/Hermes/Readme.md (hunk 3 above)
Comment (kb1lqd) [reply Applications/Hermes/Readme.md]: Fixed, shows the .uppers() are working!

File: Applications/Hermes/Readme.md (hunk 4 above)
Comment (kb1lqd) [reply Applications/Hermes/Readme.md]: Removed.

File: Applications/Hermes/Readme.md (hunk 3 above)
Comment (kb1lqd) [reply Applications/Hermes/Readme.md]: Actually completely removed per example further in file and to remove redundant example.

File: Applications/Hermes/Readme.md (hunk 5 above)
Comment (kb1lqd) [reply Applications/Hermes/Readme.md:33]: Prior example removed.

File: Applications/Hermes/Readme.md (hunk 6 above)
Comment (kb1lqd) [reply Applications/Hermes/Readme.md]: Removed "to"

File: Applications/Hermes/hermesflaskserver.py (hunk 7 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py:11]: Push for another update to add logger.

File: Applications/Hermes/hermesflaskserver.py (hunk 8 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Updated.

File: Applications/Hermes/hermesflaskserver.py (hunk 9 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Directly used config call.

File: Applications/Hermes/hermesflaskserver.py (hunk 10 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Updated to getint(). Needed to convert to string when concatenating in main `unitname = unitcallsign + '-' + str(unitnodeid)` line 169.

File: Applications/Hermes/hermesflaskserver.py (hunk 11 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: You are correct good sir. Now just returns local.

`
def configparse():
    """ Parses configuration file for callsigns and node ID and returns them in JSON format"""
    local = {}
    units = range(0, config.getint('PROXY', 'UNITS'))

    for unit in units:
        # TODO We don't really check for valid input here yet
        item = "UNIT" + str(unit)
        callsign = config.get(item, "callsign").upper()
        nodeid = config.getint(item, "nodeid")

        local[str(item)] = \
            {
                "callsign": callsign,
                "nodeid": nodeid,
            }

    return local
`

File: Applications/Hermes/hermesflaskserver.py (hunk 12 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py:70]: Updated:

` :return: Returns HTTP status along with JSON element containing the next received data from the receiver queue`

File: Applications/Hermes/hermesflaskserver.py (hunk 15 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Correct, removed.

File: Applications/Hermes/hermesflaskserver.py (hunk 13 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Removed.

File: Applications/Hermes/hermesflaskserver.py (hunk 14 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: You get a KeyError. I tried it. I updated with exception handling below, will update to logging later.

``` python
except KeyError as e:
     print "KeyError:", e

except:
     print "Unexpected Error:", sys.exc_info()[0]
```

Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Awesome, yes this is better. I don't want to have to pull in APRS.py so I just copied the function. We may want to pull this into a general functions module...

I updated several instances in the remaining code that also needed this function.

File: Applications/Hermes/hermesflaskserver.py (hunk 19 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Fixed

File: Applications/Hermes/hermesflaskserver.py (hunk 20 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Updated from prior instance.

File: Applications/Hermes/hermesflaskserver.py (hunk 21 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Removed all needs for Cpickle and used only JSON.

File: Applications/Hermes/hermesflaskserver.py (hunk 22 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: No, queue is more descriptive of what is intended. Maybe "queuelen"

Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Updated to '/queuelen'

File: Applications/Hermes/hermesflaskserver.py (hunk 16 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: Removed extra variables and just used hermes...

File: Applications/Hermes/hermesflaskserver.py (hunk 17 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskserver.py]: No, key is the key item as a string:

image


image

File: Applications/Hermes/hermesobject.py (hunk 23 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesobject.py:22]: ? I don't understand your comment.

image

File: Applications/Hermes/hermesobject.py (hunk 24 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesobject.py]: These are things you really don't want to mess with unless you really know what you're doing and go change a lot of other code/firmware.

I created global CONSTANTS:

image

File: Applications/Hermes/hermesobject.py (hunk 26 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesobject.py:74]: Good catch, added.

File: Applications/Hermes/hermesobject.py (hunk 25 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesobject.py]: ? or I already fixed and forgot I did:

image

File: Applications/Hermes/hermesflaskexample_rx.py (hunk 27 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskexample_rx.py:10]: Same reponse: Push to another update to add logging.

https://github.com/FaradayRF/Faraday-Software/pull/133#discussion_r108609659

File: Applications/Hermes/hermesflaskexample_rx.py (hunk 28 above)
Comment (kb1lqd) [reply Applications/Hermes/hermesflaskexample_rx.py:17]: Cpickle use eliminated.

**Comment by kb1lqd** [comment]
@kb1lqc Review items addressed. I realize that I just dismissed your review and that was probably the wrong way to say I'm done implemented :(

**Comment by kb1lqd** [comment]
Travis CI says there is no rakefile?

https://travis-ci.org/FaradayRF/Faraday-Software/builds/217988531

**Comment by reillyeon** [comment]
This branch doesn't have a `.travis.yml` file so it's assuming the project is Ruby and using a default configuration.

**Comment by kb1lqc** [comment]
OK testing now on hardware. Will likely suggest we make the documentation a little cleaner i.e. not immediately obvious how to setup Proxy for two units (maybe refer to proxy page?). Shouldn't hold this back but would be a nice TODO for a beginner issue ticket

**Review by kb1lqc — APPROVED** [review]
Approving but I think two or more issue tickets should be made to force some updates:
- Better document proxy setup for two units and clearly indicate that here
- Better document/show (licap?) that proxy will scroll through a bunch of errors/warnings in a nominal operation of this application as it currently is. Explain breifly.
- Some of the try/except  statements try to cram a lot of different code into the same try statement.
- Logging should be used instead of printing where possible.

@kb1lqd

**Pull request merged**